Improved UI

DiabeticOwl opened the pull request:
Using [bubbles's pogressbar](https://github.com/charmbracelet/bubbles) the section of CPU health can display each thread's usage with a progress bar.

DiabeticOwl committed:
Initial UI and ticker. Only Cpu Usage shown.

DiabeticOwl committed:
Initial display of User's computer healt statistics.

DiabeticOwl committed:
Cpu Dynamic Table.

DiabeticOwl committed:
Added progress bar to the CPU health.

Using [bubbles's pogressbar](https://github.com/charmbracelet/bubbles) the section of CPU health can display each thread's usage with a progress bar.

DiabeticOwl committed:
Improvements in Cpu Usage Table.

The progress bars takes a lot of CPU usage. Considering an alternative display/widget is advised.

DiabeticOwl committed:
Main UI completed.

CPU Usage, Memory; Disk and Processes info are now displayed in tables. Progresses bar's animation removed for performance improvements.

DiabeticOwl committed:
Main UI completed.

Progress bar planned to be moved to another PR.

DiabeticOwl committed:
Merge remote-tracking branch 'origin/htop-clone-UI' into htop-clone-UI

Fixing commit.

DiabeticOwl committed:
Fixed bug.

Fixed bug found due to commit management.

DiabeticOwl committed:
UI changes.

Changes like color and spacing between tables.

DiabeticOwl committed:
Full Screen mode implemented.

With full screen mode implemented the application is now a bit more responsive. Resizing of the tables and its contents are still needed according window's size.

DiabeticOwl committed:
Comments cleansing.

DiabeticOwl committed:
Full Screen mode upgraded.

Added versatility to the UI when resizing.

JonahBraun commented on main.go:
```diff
@@ -1,58 +1,14 @@
+// htop-clone runs a bubbletea application that displays the user's computer
```
it actually displays metrics, statistics are metrics over time or across many nodes

JonahBraun commented on main.go:
```diff
@@ -1,58 +1,14 @@
-		}
+	p := tea.NewProgram(NewModel(), tea.WithAltScreen())
+	if err := p.Start(); err != nil {
+		panic(err)
```
do you know what errors will be seen here?

DiabeticOwl committed:
Removed Niceness and Added new file system.

Removed niceness from the processes table and added the "APFS" file system for MAC users.

DiabeticOwl committed:
Added profiling capabilites and some code management.

Now profiling is available through the "cpuProfile" tag. Changed the map logic for the system's information into a struct logic.

DiabeticOwl committed:
Improved window responsiveness.

By adding pagination to disks and comments on how to navagiate through both the disks and processes tables the responsiveness of the program's window is slightly improved.

DiabeticOwl committed:
UI responsiveness upgraded and comments on file.

The UI responsiveness was enhanced by taking into consideration a set of heights, toggling tables accordingly.

DiabeticOwl committed:
Updated comment.

DiabeticOwl committed:
Enabling the catch of SIGTERM signal.

DiabeticOwl committed:
Performance improvements and new sorting mechanism.

Linux and Darwin based builds now have a significant improvement in the extraction of processes information by replacing the usage of the gopsutil/v3/process package with the execution of the ps command. Windows builds use the aforementioned package. The new sorting mecanism comes in Evertras/bubble-table==0.14.6

DiabeticOwl committed:
Removed unnecesary comments.

DiabeticOwl committed:
Comments on the usage of the ps command.

DiabeticOwl committed:
Removed unnecesary format.

DiabeticOwl committed:
Fixed bug in darwin based systems.

A bug was found due to the version of the ps commands on tested Macbook Pro.

DiabeticOwl committed:
Processes table left aligned and column selection for Darwin based systems.

DiabeticOwl committed:
Application's optimization documentation and design updates.

DiabeticOwl committed:
Go version updated.

DiabeticOwl committed:
Application's optimization documentation updated.

DiabeticOwl committed:
Documentation fix.

DiabeticOwl committed:
UI Changes and updated README preview.

JonahBraun commented on main.go:
```diff
@@ -1,58 +1,35 @@
+// Tag for activating Go's profiling tools.
+// The given value will be used as the name of the file in which the profile
+// will be saved.
+var cpuProfile = flag.String("cpuProfile", "", "Writes CPU profile to a file.")
```
since this is the only flag, and you use it only once, I recommend putting it in your function rather than a global

JonahBraun replied:
or just take it out since profiling is a one off thing and not something your users will be doing

DiabeticOwl replied:
Removed as not needed anymore. Thanks.

JonahBraun commented on processes_darwin.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+const (
+	// Columns width.
```
improve comment to explain what is going on and how it is used

DiabeticOwl replied:
Comment improved. Please let me know if the comment in [here](https://github.com/DiabeticOwl/htop-clone/pull/2/commits/bb66c2ffd63660ca2f3a962d4caad30f5f4397ab#diff-02c2fc23708d664f24ba0f1f93bd29a9b692f04206a783d17ef73d8020f16dd0) is descriptive enough.

JonahBraun commented on processes_darwin.go:
```diff
@@ -0,0 +1,73 @@
+	hugeW  = largeW + largeW
+)
+
+func extractProcessesInfo() []processInfo {
```
add a brief function comment explaining what is different in darwin/liniux

DiabeticOwl replied:
Done [here](https://github.com/DiabeticOwl/htop-clone/pull/2/commits/0356ba735ea72bc6629c890f70dc42fb7916f902). Let me know if it is clear enough.

JonahBraun commented on stats.go:
```diff
@@ -0,0 +1,104 @@
+)
+
+var (
+	desiredFileSystems = map[string]struct{}{
```
`fsFilter` might be a better var name for this

JonahBraun replied:
add comment:

```
// Display these filesystems only, ignoring uninteresting ones like <insert names here>
```

JonahBraun commented on stats.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+// extractMemoryInfo returns virtual and swap memory.
+func extractMemoryInfo() (memoryInfo, memoryInfo) {
```
`getMemInfo()` instead of `extract...` makes more sense for these names. Applies to all funcs in this file.

JonahBraun commented on ui-tables.go:
```diff
@@ -0,0 +1,211 @@
+func generateCpuTableRows(m model) []table.Row {
+	cpuFmt := "CPU #%02d:"
+	// valStr := fmt.Sprintf("%-1s", " CPU #%02d: ") + "%s"
+	rowCount := int(math.Ceil(float64(len(m.CpuInfo)) / cpuTableMaxColumnAmount))
```
a comment would be nice here

JonahBraun commented on ui-tables.go:
```diff
@@ -0,0 +1,211 @@
+	// valStr := fmt.Sprintf("%-1s", " CPU #%02d: ") + "%s"
+	rowCount := int(math.Ceil(float64(len(m.CpuInfo)) / cpuTableMaxColumnAmount))
+
+	if rowCount*cpuTableMaxColumnAmount != len(m.CpuInfo) {
```
a comment explaining under what conditions (theoretical or not) this would occur.

Alternatively, remove the if and just panic.

JonahBraun replied:
Should add this comment:

```
 	// TODO: Support core counts that are not divisible by 4.
	//       Should dynamically calculate cols from terminal width.
	//       It should also be ok if the last row is only partially filled.
```

JonahBraun commented on README.md:
```diff
@@ -1,4 +1,7 @@
+
+htop-clone is a simplified version of the [htop](https://htop.dev/) command in Linux.
+
+A description of the optimization process done over this application can be found [here](appOptimization.md).
```
Either here in your readme, or in main.go as a package comment, list out the files with a short comment as to what is in them.

Basically write a code guide for readers.

DiabeticOwl committed:
Comments added.

DiabeticOwl committed:
Comments added and README updated.

DiabeticOwl committed:
Comments Fix and Dependencies upgraded.

DiabeticOwl committed:
*.md files code organization.

JonahBraun commented on ui-tables.go:
```diff
@@ -0,0 +1,240 @@
+	// len(m.CpuInfo) / rowCount = cpuTableMaxColumnAmount, the program can't
+	// display all cores properly.
+	if rowCount*cpuTableMaxColumnAmount != len(m.CpuInfo) {
+		s := "The amount of columns or rows is incorrect.\n"
```
`s := "TODO: Support core counts that are not divisible by 4.\n"`

JonahBraun commented on ui-tables.go:
```diff
@@ -0,0 +1,240 @@
+		panic(s)
+	}
+
+	// Since each bubble-table row needs to be returned with all the information
+	// once we need to populate each with the information of the corresponding
+	// columns. For example, using a 12 core CPU and 4 columns each row:
```
```suggestion
	// Each bubble-tea row is a string containing all the contents of the columns.
	// For example, using a 12 core CPU and 4 columns each row:
```

DiabeticOwl committed:
Comment improved at processes_darwin.go

DiabeticOwl committed:
Update ui-tables.go

Co-authored-by: Jonah Braun <Jonah@goodcoffee.ca>

DiabeticOwl committed:
Difference between Darwin and Linux based systems in the 'getProcessesInfo' function  commented.

DiabeticOwl committed:
Other fixes in PR comments.

JonahBraun reviewed (approved):
🚀 This looks great! Merge it!
